crates/optimism/flashblocks/src/consensus.rs

```diff
@@ -1,86 +1,143 @@
+                        "error while fetching flashblock completed sequence",
+                    );
+                    continue;
+                }
+            };
 
-                    if sequence.state_root().is_none() {
-                        warn!(target: "flashblocks", "Missing state root for the complete sequence")
-                    }
+            let payload = match P::ExecutionData::try_from(&sequence) {
+                Ok(payload) => payload,
+                Err(err) => {
+                    error!(
+                        target: "flashblocks",
+                        %err,
+                        "error while converting to payload from completed sequence",
+                    );
+                    continue;
+                }
+            };
 
-                    // Load previous block hashes. We're using (head - 32) and (head - 64) as the
-                    // safe and finalized block hashes.
-                    let safe_block_hash = self.get_previous_block_hash(&previous_block_hashes, 32);
-                    let finalized_block_hash =
-                        self.get_previous_block_hash(&previous_block_hashes, 64);
+            let block_number = payload.block_number();
+            let block_hash = payload.block_hash();
+            // Call engine_newPayload
+            match self.engine_handle.new_payload(payload).await {
+                Ok(result) => {
+                    debug!(
+                        target: "flashblocks",
+                        flashblock_count = sequence.count(),
+                        block_number,
+                        %block_hash,
+                        ?result,
+                        "Submitted engine_newPayload",
+                    );
 
-                    let state = alloy_rpc_types_engine::ForkchoiceState {
-                        head_block_hash: block_hash,
-                        safe_block_hash,
-                        finalized_block_hash,
-                    };
+                    if result.status != PayloadStatusEnum::Valid {
+                        continue;
+                    }
+                }
+                Err(err) => {
+                    error!(
+                        target: "flashblocks",
+                        %err,
+                        block_number,
+                        "Failed to submit new payload",
+                    );
+                    continue;
+                }
+            }
 
-                    // Send FCU
-                    let _ = self
-                        .engine_handle
-                        .fork_choice_updated(state, None, EngineApiMessageVersion::V3)
-                        .await;
+            // FCU to update tip, only after successful engine_newPayload
+            let fcu_state = alloy_rpc_types_engine::ForkchoiceState {
+                head_block_hash: block_hash,
+                safe_block_hash: B256::ZERO,
+                finalized_block_hash: B256::ZERO,
```

Comment:
this works once the node caught up, but if you want to sync a new node via flashblocks then we also need to choose a finalized block so we can kick off p2p backfill sync

I think we could simply use the parent block here for now?

ideally we track the some recent blocks based on 32slots*6secs blocks here and assume that new blocks are being built on the canonical chain